common/numerated/src/numerated.rs

```diff
@@ -0,0 +1,181 @@
+    Upper(T),
+}
+
+/// For any type `T`, `Bound<T>` is a type, which has set of values bigger than `T` by one element.
+/// - Each value from `T` has unambiguous mapping to `Bound<T>`.
+/// - Each value from `Bound<T>`, except one called __upper__, has unambiguous mapping to `T`.
+/// - __upper__ value has no mapping to `T`, but can be used to get `T` max value.
+///
+/// # Examples
+/// 1) For any `T`, which max value can be get by calling some static live time function,
+/// Option<T> can be used as `Bound<T>`. `None` is __upper__. Mapping: Some(t) -> t, t -> Some(t).
+///
+/// 2) When `inner` field max value is always smaller than `inner` type max value, then we can use this variant:
+/// ```
+/// use numerated::{Bound, BoundValue};
+///
+/// /// `inner` is a value from 0 to 99.
+/// struct T { inner: u32 }
+///
+/// /// `inner` is a value from 0 to 100.
+/// #[derive(Clone, Copy)]
+/// struct B { inner: u32 }
+///
+/// impl From<T> for B {
+///     fn from(t: T) -> Self {
+///         Self { inner: t.inner }
+///     }
+/// }
+///
+/// impl Bound<T> for B {
+///    fn unbound(self) -> BoundValue<T> {
+///        if self.inner == 100 {
+///            BoundValue::Upper(T { inner: 99 })
+///        } else {
+///            BoundValue::Value(T { inner: self.inner })
+///        }
+///    }
+/// }
+/// ```
+pub trait Bound<T: Sized>: From<T> + Copy {
+    /// Unbound means mapping bound back to value if possible.
+    /// - In case bound is __upper__, then returns Upper(max), where `max` is `T` max value.
+    /// - Otherwise returns Value(value).
+    fn unbound(self) -> BoundValue<T>;
+    fn get(self) -> Option<T> {
+        match self.unbound() {
+            BoundValue::Value(v) => Some(v),
+            BoundValue::Upper(_) => None,
+        }
+    }
+}
+
+/// Numerated type is a type, which has type for distances between any two values of `Self`,
+/// and provide an interface to add/subtract distance to/from value.
+pub trait Numerated: Copy + Sized + Ord + Eq {
+    /// Numerate type: type that describes the distances between two values of `Self`.
+    type N: PrimInt + Unsigned;
+    /// Bound type: type for which any value can be mapped to `Self`,
+    /// and also has __upper__ value, which is bigger than any value of `Self`.
+    type B: Bound<Self>;
+    /// Adds `num` to `self` if `self + num` is between `self` and `other`.
+    fn add_if_between(self, num: Self::N, other: Self) -> Option<Self>;
+    /// Subtracts `num` from `self` if `self - num` is between `self` and `other`.
+    fn sub_if_between(self, num: Self::N, other: Self) -> Option<Self>;
+    /// Returns `self - other` if `self >= other`.
+    fn distance(self, other: Self) -> Option<Self::N>;
+    /// Increments `self` if `self < other`.
+    fn inc_if_lt(self, other: Self) -> Option<Self> {
+        self.add_if_between(Self::N::one(), other)
+    }
+    /// Decrements `self` if `self > other`.
+    fn dec_if_gt(self, other: Self) -> Option<Self> {
```

Comment:
For rust math crates and `num-traits` it's common way to use names like `add`, `sub`, `inc`, `dec`, `lt`, `gt`. Because this crate is also positioning as math I think it's better to use this common names